update-python-libraries: Add support for fetchgit, package can deny bulk updates, and more robust hash matching

mweinelt opened the pull request:
Previously when packages that required the git fetcher were updated, we would wrongly rely on `nix-prefetch-url`, which would reliable break the hash.

Instead we need to use `nix-prefetch-git` to determine the proper hash, when the relevant attributes are present.

Implements the blocklist for certain packages, that was requested on matrix.

And makes hash replacement more robust, by parsing and reusing the `outputHash` for matching the current line and value.

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

mweinelt committed:
update-python-libraries: Add support for fetchgit

Previously when packages that required the git fetcher were updated, we
would wrongly rely on `nix-prefetch-url`, which would reliable break the
hash.

Instead we need to use `nix-prefetch-git` to determine the proper hash,
when the relevant attributes are present.

mweinelt committed:
update-python-libraries: Allow packages to deny bulk updates

By settings `passthru.skipBulkUpdate = true` in a python package, it
will be excluded from bulk update runs.

This is useful in case of fragile packages, that require special
attention during updates.

mweinelt committed:
update-python-libraries: Make the hash replacement more robust

Retrieve the current package hash and use it for a more precise match,
which allows us to update a derivation in situation where multiple
hashes are present.

SuperSandro2000 commented on maintainers/scripts/update-python-libraries:
```diff
@@ -1,5 +1,3 @@
-exec ${python} pkgs/development/interpreters/python/update-python-libraries/update-python-libraries.py $@
-
+#!/usr/bin/env nix-shell
+#!nix-shell -I nixpkgs=channel:nixpkgs-unstable -i bash -p "python3.withPackages (ps: with ps; [ packaging requests ])" -p nix-prefetch-git
```
Can we rely on nixpkgs from NIX_PATH? I don't have any channels on my systems so this would fail, right?

mweinelt replied:
Wby don't you give that command line a try and tell me?

SuperSandro2000 replied:
I wasn't on a PC when I wrote this but turns out it works.

FRidh commented on pkgs/development/interpreters/python/update-python-libraries/update-python-libraries.py:
```diff
@@ -20,9 +21,26 @@
 import collections
 import subprocess
 
+BLOCKLIST = {
```
In the past I wanted to introduce a `passthru` attribute that indicated the updating strategy. That also gave the maintainers a way to say "Don't update my package". Not saying we should go that way, but it is an alternative.

mweinelt replied:
That sounds like a better approach, going to look into it.

mweinelt replied:
I decided for `passthru.denyBulkUpdate`. Unless anyone has a better name in mind.

mweinelt replied:
```diff
diff --git a/pkgs/development/python-modules/wandb/default.nix b/pkgs/development/python-modules/wandb/default.nix
index e1139ecf1b1..8764eca6ab9 100644
--- a/pkgs/development/python-modules/wandb/default.nix
+++ b/pkgs/development/python-modules/wandb/default.nix
@@ -164,6 +164,8 @@ buildPythonPackage rec {
     "wandb"
   ];
 
+  passthru.denyBulkUpdate = true;
+
   meta = with lib; {
     description = "A CLI and library for interacting with the Weights and Biases API";
     homepage = "https://github.com/wandb/wandb";
```

```
❯ maintainers/scripts/update-python-libraries --target minor --commit --use-pkgs-prefix pkgs/development/python-modules/wandb/default.nix
INFO:root:Updating packages...
WARNING:root:Path /home/hexa/git/nixos/python-updater/pkgs/development/python-modules/wandb/default.nix: Bulk update denied for wandb
INFO:root:Finished updating packages.
INFO:root:Using python3Packages. prefix for commits
INFO:root:Committing updates...
INFO:root:Finished committing updates
INFO:root:0 package(s) updated
```

FRidh replied:
Do we want to limit it only to our bulk updates or also r-ryantm updates? Judging from the name it seems to be the former, and I think that's generally speaking also what is desired.

mweinelt replied:
I think allowing updates through r-ryantm is fine, since they go through a dedicated review. Not our business really.

mweinelt replied:
Not sure how to distinguish between the two though.

mweinelt replied:
If more than one package is affected, I now toggle a global knob. Not pretty, but it does the trick.

FRidh commented on pkgs/development/interpreters/python/update-python-libraries/update-python-libraries.py:
```diff
@@ -67,6 +85,21 @@ def _get_values(attribute, text):
+def get_attr_value(attr_path: str) -> Optional[Any]:
+    try:
+        response = subprocess.check_output([
+            "nix",
```
Maybe enable the experimental feature like is done further down below.

mweinelt replied:
Added.

FRidh commented on pkgs/development/interpreters/python/update-python-libraries/update-python-libraries.py:
```diff
@@ -344,12 +411,18 @@ def _update_package(path, target):
     # sri hashes have been the default format since nix 2.4+
     sri_hash = subprocess.check_output(["nix", "--extra-experimental-features", "nix-command", "hash", "to-sri", "--type", "sha256", new_sha256]).decode('utf-8').strip()
```
Further up (~282) we call the same command. Factor out?

mweinelt replied:
Extracted into `_hash_to_sri` function.

mweinelt committed:
doc/python: Explain opt-out out of bulk updates

mweinelt committed:
update-python-libraries: Improve code quality

- Prune unused imports
- Collect imports at the top
- Removed unused exception assignments
- Fill bare except clause with
- Expand overly long check_output lines

mweinelt commented:
With these changes I get 31 more (431 -> 460) packages updated then without.

mweinelt commented:
- Spotted an issue updating `ibis-framework`, where it would update the wrong fetcher. 52d2e3c4ada4a05239db64069710be59cf3ae912
